### DISCUSSION TITLE
Question - any particular reason why `Debug` trait isn't derived for the `Value` enum?

### DISCUSSION THREAD

**Discussion by nihalpasham** [tock/tock#2931] [Q&A] [answered]
I was wondering if there's a specific reason - why `Debug` is not derived for `Value`. I think it can help with debugging in a couple of  scenarios.

For example, I was debugging a driver and needed to print-out some bitfields but had to put together my own `Debug` impl for quick debugging. I guess, if someone wants to customize their prints, they could always implement `Display` while `Debug` serves as the most basic log format.

```rust
impl Debug for SCR::BUS_WIDTH::Value {
    fn fmt(&self, _f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::BUS_WIDTH_1 => {
                print!("WIDTH_1");
                Ok(())
            }
            Self::BUS_WIDTH_4 => {
                print!("WIDTH_4");
                Ok(())
            }
            Self::BUS_WIDTH_1_4 => {
                print!("WIDTH_1_4");
                Ok(())
            }
        }
    }
}

```

**Comment by hudson-ayers** [comment, marked as answer]
The main reason to avoid using `#[derive(Debug)]` is that Rust's formatting machinery is pretty heavy and relying on it to print much leads to binaries with a lot of bloat as a result (see https://github.com/tock/tock/blob/master/doc/CodeSize.md#formatting-overhead and https://jamesmunns.com/blog/fmt-unreasonably-expensive/ for discussions of this). Of course, by simply not printing the value in question you avoid the code size overhead, so we could derive `Debug` without hurting code size in the common case -- it just lowers the barrier to more widespread use of `Debug` which can ultimately lead to code bloat.

**Comment by nihalpasham** [comment]
`debug prints` are technically zero-cost (i.e. we don't pay for what we don't use) and deriving it would make for easier `debug-builds` but I guess, if the goal is to help drive a certain user-behavior then this makes sense.

**Reply by hudson-ayers** [reply]
Yeah, I don't feel super strongly about this. I could be convinced that we should `#[derive(Debug)]` and just warn users about the potential bloat and try to monitor for people printing these values in code that makes it upstream. If there was an easy way to check in CI that none of our upstream boards print these values I would be all for it.
